util/merkle-mountain-range/src/mmr.rs

```diff
@@ -0,0 +1,196 @@
+//! Merkle Mountain Range
+//!
+//! references:
+//! https://github.com/mimblewimble/grin/blob/master/doc/mmr.md#structure
+//! https://github.com/mimblewimble/grin/blob/0ff6763ee64e5a14e70ddd4642b99789a1648a32/core/src/core/pmmr.rs#L606
+
+use crate::helper::{get_peaks, parent_offset, pos_height_in_tree, sibling_offset};
+use crate::mmr_store::{MMRBatch, MMRStore};
+use crate::MerkleElem;
+use crate::{Error, Result};
+use std::borrow::Cow;
+use std::convert::TryInto;
+use std::fmt::Debug;
+use std::marker::PhantomData;
+
+pub struct MMR<'a, Elem: MerkleElem, Store: MMRStore<Elem>> {
```

Comment:
we should define a `Merge` trait with an associated type, which can resolve orphan rules issue and simplifies generic type definitions of related structs.

```
pub trait Merge {
    type Item;
    fn merge(left: &Self::Item, right: &Self::Item) -> Self::Item;
}

pub struct MMR<'a, T, M, S: MMRStore<T>> {
    mmr_size: u64,
    batch: &'a mut MMRBatch<T, S>,
    merge: PhantomData<M>,
}
```

Reply:
I can't see the advantage compare with the current code, it's just two different styles.

Reply:
`merkle_elem: PhantomData<Elem>,` is useless, we can remove it.